### PR TITLE
Feature/integration specs and fix n+1 problems

### PR DESCRIPTION
**Pull request by Salimer**
### n+1

- Make sure that the N+1 problem is solved when fetching all posts and their comments for a user.
    - Do not use the bullet gem in this exercise - instead analyze the logs as showed in this [article](https://dev.to/junko911/rails-n-1-queries-and-eager-loading-10eh)
    - Include the screenshots of your app's logs before and after the fix

### before
before

### after

after

### before
Screenshot 2023-08-29 at 10 50 09

### after
Screenshot 2023-08-29 at 10 50 30

### Integration specs

- Use Capybara to write integration tests for each view in your project.

- Think on the flow of things and group the tests of page in a way that makes sense. For example, one test would check for the things in the page and another test for clicking links on it.

- User index page:

    - I can see the username of all other users.
    - I can see the profile picture for each user.
    - I can see the number of posts each user has written.
    - When I click on a user, I am redirected to that user's show page.

- user show page:

    - I can see the user's profile picture.
    - I can see the user's username.
    - I can see the number of posts the user has written.
    - I can see the user's bio.
    - I can see the user's first 3 posts.
    - I can see a button that lets me view all of a user's posts.    
    - When I click a user's post, it redirects me to that post's show page.    
    - When I click to see all posts, it redirects me to the user's post's index page.

- User post index page:
 
    - I can see the user's profile picture.
    - I can see the user's username.
    - I can see the number of posts the user has written.
    - I can see a post's title.
    - I can see some of the post's body.
    - I can see the first comments on a post.
    - I can see how many comments a post has.
    - I can see how many likes a post has.
    - I can see a section for pagination if there are more posts than fit on the view.
    - When I click on a post, it redirects me to that post's show page.

- Post show page:

    - I can see the post's title.
    - I can see who wrote the post.
    - I can see how many comments it has.
    - I can see how many likes it has.
    - I can see the post body.
    - I can see the username of each commentor.
    - I can see the comment each commentor left.

### CODE REVIEW

File: app/models/comment.rb [hunk 1]
```diff
@@ -7,11 +7,15 @@ class Comment < ApplicationRecord
...
-  def update_post_comments_counter
-    post.update(comments_counter: post.comments.count)
+  def increase_post_comments_counter
+    post.increment!(:comments_counter)
+  end
+
+  def decrement_post_comments_counter
+    post.decrement!(:comments_counter)
   end
 end
```
Comment (Gambit142) [inline app/models/comment.rb:21]: - [ ] The following change should be done in the `index` action of the `Posts` Controller and anywhere that needs the **_N+1 query_** corrected:

- Kindly correct the **_N+1 query_** in the controller. Also, the screenshot that you took should be that of the terminal logs before correcting the **_N+1 query_** and after correcting the **_N+1 query_**. Kindly fix this and attach the correct screenshot as it is part of this milestone's requirements. 👍 🖱️

**Review by Gambit142 — CHANGES_REQUESTED** [review]
Hi @Salimer  @AliShahSafdari 👋, 

Good job so far!

## Highlights
- Setup linters correctly ✔️ 
- Wrote features test for the `Posts#Index` and the `Posts#Show` pages ✔️ 
- Wrote features test for the `Posts#Index` and the `Posts#Show` pages ✔️

You still need to work on some issues to go to the next project, but you are almost there!

## Required Changes ♻️
_Kindly check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend considering them as they can improve your code._

You can also consider the following:

N/A

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

#### Please tag me in your question so I can receive the notification. You can tag me by typing @Gambit142

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
```

File: app/models/comment.rb (hunk 1 above)
Comment (Salimer) [reply app/models/comment.rb:21]: I don't really understand how or why should I add callbacks to the controller. However, we updated the controllers and views to fairly improve the N+1 problems

**Review by OLIPLICHE — CHANGES_REQUESTED** [review]
Hi Team, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
- Work in pair programming 👍 
- Has Capybara gem in Gemfile 👍 
- Has integration test 👍 

 ## Required Changes ♻️

- [ ] [Here is an example of how to solve N+1 problem](https://blog.devgenius.io/avoiding-the-n-1-query-problem-in-rails-2f571f9e57e5): In the posts#index actions, you should use `@posts = @user.posts.includes(:comments)` instead of `@posts = @user.posts`

Capture
- NB: Kindly pay attention to the red sign in the picture, which is your code to understand what I am talking about.
- This comment is not in line with the code because you did not make the concerned file visible in this PR

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Salimer** [comment]
Hello @OLIPLICHE  and thank you for your review

I tried to add this line `@posts = @user.posts.includes(:comments)` instead of `@posts = @user.posts`, but actually it just adds one query to the logs. Since I am calling the `five_most_recent_comments` method with each post then it makes new query nonetheless.
### Here is the terminal log with `@posts = @user.posts.includes(:comments)`
Screenshot 2023-08-29 at 09 26 26

### Here is with `@posts = @user.posts`

Screenshot 2023-08-29 at 09 25 54


With the requested change I am requesting 8 queries while without it I request 7. As far as I understand, trying to solve the N+1 problems means I request less queries, right?

**Comment by medaminedev66** [comment]
Hello @Salimer 👋,

_This is an additional review as a result of the second opinion request._

We can observe that there is no visible change in the screenshot even after implementing `@posts = @user.posts.includes(:comments)`. The reason we aren't experiencing a significant reduction in queries lies in the fact that we might not be encountering any N+1 problems.

This is because our current context involves accessing the `'/users/1/posts'` route, which implies we are fetching posts only for a specific user, not multiple users simultaneously. If we were, for instance, viewing a feed containing posts from different users, utilizing includes would have a more pronounced effect, leading to an immediate reduction in queries.

Let me elaborate on why we recommend implementing this optimization. When we use `includes(:comments)`, we are directly loading all the posts and their associated comments for a user. In this context, we aim to retrieve a collection of data comprising posts, and our intention is to expedite query execution by loading comments alongside posts during the data retrieval process. To illustrate this concept, consider the analogy of grocery shopping: you buy apples but you also need to get mangoes as well. Instead of having to go back to fetch mangoes later, you include them together with the apples making it take shorter.

However, it's important to understand that Microverse's intention in requiring this practice is to ensure that developers not only address immediate N+1 concerns but also build awareness of the issue from the outset. This empowers developers to tackle similar performance challenges that may arise in the future.

As this was the only reason that your project was not approved it will be approved as soon as possible.
Please note that since we will approve it manually so this could take a few hours!

Keep rocking 🚀!

**Pull request merged**